### PR TITLE
Explication de base sur les erreurs

### PR DESCRIPTION
**Pull request by Pismice**
Je ne vais pas très loin dans le système d'erreurs car la doc me parait assez complète, mais je pense que avec ça un utilisateur nouveau à ZIG peut facilement comprendre comment les erreurs fonctionnent et comment les utiliser rapidement.

J'ai également errdefer qui est spécifique à ZIG car je pense que ça peut être assez utile.

Dis moi si ça te parait clair ou si tu vois une douille :)

### CODE REVIEW

File: doc/errors.org [hunk 1]
```diff
@@ -0,0 +1,95 @@
+* Errors
+https://ziglang.org/documentation/0.10.1/#Errors
```
Comment (samuel-schlaeppy) [inline doc/errors.org]: The last version of Zig is 0.11, is it voluntary to have the 0.10 one?

File: doc/errors.org [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+In ZIG errors are juste values.
+Those values are those of an special type of enum "error".
+When you declare your enum using this keyword instead of "enum", the compiler is going to know that those values are going to be errors, therefore they can be catched, tried, ...
+
+
+** Handling errors in the application flow
+Functions can return either a value or an error, this is done by adding the "!" to the return type of a function
+With the "!" the return type of the function is the coerced "anyerror!u64"
```
Comment (samuel-schlaeppy) [inline doc/errors.org]: In the 0.11 zig doc, i found that error are coerced to `u16`, is this specific ?

File: doc/errors.org [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    try std.testing.expect(err == 3);
+}
+
+test "should panic" {
+    _ = maybeErrorMaybeValue(true) catch unreachable;
+}
+#+end_src
+
+** Errdefer
+This exemple is taken from: https://ziglang.org/documentation/0.10.1/#errdefer
```
Comment (samuel-schlaeppy) [inline doc/errors.org]: Doc is for 0.10

File: doc/errors.org [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+Errdefer is a particularity of ZIG, it allows the user to execute some code when the function returns an error, it is useful exemple for deallocating variables you would have normally returned from the function, but since the function failed you deallocate that memory to avoid a memory leak.
+#+source: simple
+#+begin_src zig :imports '(std)
```
Comment (samuel-schlaeppy) [inline doc/errors.org:46]: Mabe in the future, we can transform the example to use org-babel to demonstrate the use case.

File: doc/zig-doc.org [hunk 5]
```diff
@@ -21,6 +21,8 @@ Write about how Zig's log module uses ~comptime~ and whether the log system will
 ** TODO Errors and compilation mode
```
Comment (samuel-schlaeppy) [inline doc/zig-doc.org]: Remove the top title, it's rewrite in the `errors.org` file

File: doc/errors.org (hunk 2 above)
Comment (Pismice) [reply doc/errors.org]: https://ziglang.org/documentation/0.11.0/#Errors
You are probably reffering to this: "The number of unique error values across the entire compilation should determine the size of the error set type. However right now it is hard coded to be a u16. See [#768](https://github.com/ziglang/zig/issues/786)."

What I meant with this is that when we add the "!" my return value becomes something like "anyerror!string", "anyerror!bool".

And yes what you said is correct from my understanding "anyerror" is simply an "u16" under the hood.

My documentation was indeed not precise, not in relation with my exemple and could lead to misenterpretation, I changed it ty :)

`With the "!" the return type of the function below is the coerced "anyerror!u8"

When a function can return an error we have to either "try" or "catch" it.

#+source: simple
#+begin_src zig :imports '(std)
fn maybeErrorMaybeValue(isError: bool) !u8 {
    // ... might return an u8 or an error
    if (isError) {
        return error.Bar;
    } else {
        return 2;
    }
}`

I don't know if saying in my doc what is "anyerror" under the hood is  useful, what do you think ?

File: doc/errors.org (hunk 3 above)
Comment (Pismice) [reply doc/errors.org]: I accidently was reading the 0.10 doc while doing this, I did all my exemples on 0.11 and just checked the 2 docs and see no difference between the 2 on this subject

File: doc/errors.org (hunk 4 above)
Comment (Pismice) [reply doc/errors.org:46]: Yes :)

File: doc/errors.org (hunk 1 above)
Comment (Pismice) [reply doc/errors.org]: same reponse as https://github.com/Pismice/HEIG_ZIG/pull/2#discussion_r1518894572

File: doc/errors.org (hunk 3 above)
Comment (samuel-schlaeppy) [reply doc/errors.org]: Ok, mabe we can change the link to 0.11

File: doc/errors.org (hunk 1 above)
Comment (samuel-schlaeppy) [reply doc/errors.org]: see: [comment change link](https://github.com/Pismice/HEIG_ZIG/pull/2#discussion_r1518905377)

File: doc/errors.org (hunk 2 above)
Comment (samuel-schlaeppy) [reply doc/errors.org]: Ok, I understand. I think that it's not necessary to include the information about the issue [#786](https://github.com/ziglang/zig/issues/786).
But we can say it more explicitly as in the standard doc [chapter : error-union-type](https://ziglang.org/documentation/0.11.0/#Error-Union-Type)

**Comment by Pismice** [comment]
Its ok for me to merge :)

**Pull request merged**